commit 284e3ca3aac527197eb513b7aa3a1269cb399994 (refs/changes/51/651/2)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2016-10-11T15:27:59-07:00 (3 years ago)
    
    MANTA-3001 electric-moray should register all of its ports in binder

diff --git a/.gitignore b/.gitignore
index b2ca767..ea64500 100644
--- a/.gitignore
+++ b/.gitignore
@@ -7,3 +7,4 @@ cscope.in.out
 cscope.po.out
 cscope.out
 smf/manifests/*.xml
+sapi_manifests/registrar/template
diff --git a/Makefile b/Makefile
index 986431e..f7486ce 100644
--- a/Makefile
+++ b/Makefile
@@ -39,6 +39,7 @@ JSSTYLE_FILES	 = $(JS_FILES)
 JSSTYLE_FLAGS    = -f tools/jsstyle.conf
 REPO_MODULES	 = src/node-dummy
 SMF_MANIFESTS_IN = smf/manifests/haproxy.xml.in
+BOOTSTRAP_MANIFESTS = sapi_manifests/registrar/template
 
 
 NODE_PREBUILT_VERSION=v0.10.25
@@ -77,7 +78,7 @@ deps: | $(REPO_DEPS) $(NPM_EXEC)
 $(NODEUNIT): | $(NPM_EXEC)
 	$(NPM) install
 
-CLEAN_FILES += $(TAP) ./node_modules
+CLEAN_FILES += $(TAP) ./node_modules $(BOOTSTRAP_MANIFESTS)
 
 .PHONY: manta-scripts
 manta-scripts: deps/manta-scripts/.git
@@ -92,7 +93,7 @@ test: $(NODEUNIT)
 	$(NODEUNIT) test/integ.test.js | $(BUNYAN)
 
 .PHONY: release
-release: all $(SMF_MANIFESTS)
+release: all $(SMF_MANIFESTS) $(BOOTSTRAP_MANIFESTS)
 	@echo "Building $(RELEASE_TARBALL)"
 	@mkdir -p $(RELSTAGEDIR)/root/opt/smartdc/electric-moray
 	@mkdir -p $(RELSTAGEDIR)/root/opt/smartdc/boot
@@ -119,6 +120,14 @@ release: all $(SMF_MANIFESTS)
 	(cd $(RELSTAGEDIR) && $(TAR) -jcf $(ROOT)/$(RELEASE_TARBALL) root)
 	@rm -rf $(RELSTAGEDIR)
 
+# We include a pre-substituted copy of the template in our built image so that
+# registrar doesn't go into maintenance on first boot. Then we ship both the
+# .in file and this "bootstrap" substituted version. The boot/setup.sh script
+# will perform this replacement again during the first boot, replacing @@PORTS@@
+# with the real ports list.
+sapi_manifests/registrar/template: sapi_manifests/registrar/template.in
+	sed -e 's/@@PORTS@@/2020/g' $< > $@
+
 .PHONY: publish
 publish: release
 	@if [[ -z "$(BITS_DIR)" ]]; then \
diff --git a/boot/setup.sh b/boot/setup.sh
index 43139a6..4d6a589 100755
--- a/boot/setup.sh
+++ b/boot/setup.sh
@@ -118,6 +118,21 @@ function manta_setup_electric_moray {
         ports[$i]=`expr 2020 + $i`
     done
 
+    #Regenerate the registrar config with the real ports included
+    #(the bootstrap one just includes 2020 alone)
+    IFS=','
+    local portlist=$(echo "${ports[*]}" | sed 's/^,//')
+    local RTPL=$SVC_ROOT/sapi_manifests/registrar/template
+    sed -e "s/@@PORTS@@/${portlist}/g" ${RTPL}.in > ${RTPL}
+
+    #Wait until config-agent regenerates config.json before restarting
+    #registrar
+    svcadm restart config-agent
+    while [[ /opt/smartdc/registrar/etc/config.json -ot ${RTPL} ]]; do
+        sleep 1
+    done
+    svcadm restart registrar
+
     #To preserve whitespace in echo commands...
     IFS='%'
 
diff --git a/sapi_manifests/registrar/template b/sapi_manifests/registrar/template.in
similarity index 94%
rename from sapi_manifests/registrar/template
rename to sapi_manifests/registrar/template.in
index b4ec45b..ecc66a0 100644
--- a/sapi_manifests/registrar/template
+++ b/sapi_manifests/registrar/template.in
@@ -11,7 +11,8 @@
       },
       "ttl": 60
     },
-    "ttl": 30
+    "ttl": 30,
+    "ports": [@@PORTS@@]
   },
 
   "zookeeper": {
